feat: Optimize Dockerfile and update setup documentation

igopalakrishna opened the pull request:
Closes #80

This PR completely revamps the project's Docker setup to improve dependency management, build efficiency, and user experience, addressing all points outlined in the issue.

### A) Dependency Management (`pyproject.toml` & `Dockerfile`)
- Consolidated all Python package installation to use **Poetry**, removing the separate `pip install` step from the `Dockerfile`.
- Moved notebook-specific packages (`jupyter`, `ipykernel`, `ipython`) into a new `[tool.poetry.group.notebooks]` group in `pyproject.toml`.
- Added `grpcio-tools` as a development dependency to ensure reliable `.proto` file compilation within the Python environment.
- Updated the `protobuf` version to `^4.25.0` to resolve a dependency conflict with `grpcio-tools`.

### B) & C) Dockerfile Optimization
- **Best Practices:** The `Dockerfile` now follows modern conventions, including:
  - Using `WORKDIR` instead of `RUN cd`.
  - Setting `SHELL ["/bin/bash", "-o", "pipefail", "-c"]` to ensure build steps fail correctly.
  - Using `curl` consistently instead of `wget`.
- **Build Efficiency:**
  - `RUN` commands have been restructured into logical layers to improve Docker's layer caching and significantly speed up subsequent builds.
  - The build process now copies only `pyproject.toml` and `poetry.lock` before installing dependencies, preventing re-installation on minor code changes.
- **Simplicity:**
  - The complex startup script (`start.sh`) and the `git clone` step have been removed in favor of a cleaner `COPY . .` and a direct `CMD` instruction.
  - The Jupyter token is now disabled in the `CMD` for a better developer experience, matching the original `Dockerfile`'s intent.

### Expanded Documentation (`docs/setup/docker.md`)
- The Docker setup guide has been completely rewritten to be clearer and more comprehensive.
- The new guide provides instructions for running the container in both foreground and detached modes.
- It now includes specific commands for executing tasks against a running container, such as running the `pytest` suite or accessing an interactive shell, as requested in the issue.

### How to Test
1. Build the image: `docker build -t sbsim .`
2. Run the container: `docker run -p 8888:8888 sbsim`
3. Access Jupyter at `http://localhost:8888`.
4. Run tests against the detached container: `docker exec <container_name> poetry run pytest`

### Known Issue
The test suite fails with a `Fatal Python error: Illegal instruction` when run inside the container on an Apple Silicon (ARM) Mac. This is a known issue related to CPU instruction sets when running TensorFlow in an emulated environment and is outside the scope of the Dockerfile optimizations. The `Dockerfile` builds successfully, the Jupyter server runs, and the `docker exec` command is correctly configured for users on native x86/amd64 systems.

igopalakrishna committed:
feat: Optimize Dockerfile and update setup documentation

ajkhabbazi commented:
@igopalakrishna @s2t2 Tried testing this PR locally (Windows + Docker Desktop). The Docker build fails at the apt install step:

E: Unable to locate package python3.10-venv
E: Unable to locate package python3.10-distutils
(exit code 100)

CI also shows a failing ubuntu-latest (Py3.11) job. Suggest switching the base image to Ubuntu 22.04 (jammy) or adjusting the Python install approach, then re-running CI.

s2t2 commented:
Thanks @ajkhabbazi for the review!

Let's use Python 3.11 in the Docker file, to match the other operating environments.

igopalakrishna committed:
Switch to python:3.11-slim base image and fix deps

- Replace Ubuntu 20.04 + deadsnakes with python:3.11-slim
- Update protobuf to ^4.25.0 (TF 2.15.1 compatible)
- Update grpcio-tools to ^1.60.0
- Regenerate protobuf files for protobuf 4.x compatibility
- Add Apple Silicon notes to Docker docs
- Regenerate poetry.lock

s2t2 commented:
Let's revert the changes to the pb2.py files, and then as long as you think this is ready, we can merge!

ajkhabbazi commented:
@igopalakrishna -- two fixes needed based on @s2t2's feedback:
 1. Switch base image to ubuntu:22.04 and replace all python3.10 references with python3.11 -- fixes the apt install failure on Ubuntu 20.04.
 2. Revert the three pb2.py files -- they're auto-generated and shouldn't be in this PR.
Happy to take over and complete this PR if you're no longer working on it.
